16-encoding-categorical-data.Rmd

```diff
@@ -1,8 +1,356 @@
+
+## Feature hashing
+
+Traditional dummy variables as described in Section \@ref(dummies) require that all of the possible categories be known to create a full set of numeric features. _Feature hashing_ methods also create dummy variables, but only consider the value of the category to assign it to a predefined pool of dummy variables. Let's look at the `Neighborhood` values in Ames again and use the `rlang::hash()` function to understand more.
+
+```{r}
+library(rlang)
+
+ames_hashed <-
+  ames_train %>%
+  mutate(Hash = map_chr(Neighborhood, hash))
+
+ames_hashed %>%
+  select(Neighborhood, Hash)
+```
+
+If we input Briardale to this hashing function, we will always get the same output. The neighborhoods in this case are called the "keys", while the outputs are the "hashes".
+
+:::rmdnote
+A hashing function takes an input of variable size and maps it to an output of fixed size. Hashing functions are commonly used in cryptography and databases.
+:::
+
+The `rlang::hash()` function generates a 128-bit hash, which means there are `r 2^128` possible hash values. This is great for some applications but doesn't help with feature hashing of **high cardinality** variables (variables with many levels). In feature hashing, the number of possible hashes is a hyperparameter and is set by the model developer through computing the modulo of the integer hashes.
+
+```{r}
+ames_hashed %>%
+  ## first make a smaller hash for integers that R can handle
+  mutate(Hash = strtoi(substr(Hash, 26, 32), base = 16L),  
+         ## now take the modulo
+         Hash = Hash %% 16) %>%
+  select(Neighborhood, Hash)
+```
+
+Now instead of the `r n_distinct(ames_train$Neighborhood)` neighborhoods in our original data or an incredibly huge number of the original hashes, we have sixteen hash values. This method is very fast and memory efficient, and it can be a good strategy when there are a large number of possible categories.
+
+We can implement feature hashing using a recipe step from the `r pkg(textrecipes)` package.
+
+:::rmdnote
+Feature hashing is useful for text data as well as high cardinality categorical data. See Section 6.7 of @Hvitfeldt2021 for a case study demonstration with text predictors.
+:::
+
+```{r}
+library(textrecipes)
+ames_hash <- 
+  recipe(Sale_Price ~ Neighborhood + Gr_Liv_Area + Year_Built + Bldg_Type + 
+           Latitude + Longitude, data = ames_train) %>%
+  step_log(Gr_Liv_Area, base = 10) %>% 
+  step_dummy_hash(Neighborhood, signed = FALSE, num_terms = 16L) %>%
+  step_dummy(all_nominal_predictors()) %>% 
+  step_interact( ~ Gr_Liv_Area:starts_with("Bldg_Type_") ) %>% 
+  step_ns(Latitude, Longitude, deg_free = 20)
+
+ames_hash
+```
+
+Feature hashing is fast and efficient but has a few downsides. For example, different category values often map to the same hash value. This is called a _collision_ or _aliasing_. How often did this happen with our neighborhoods in Ames?
+
+```{r echo=FALSE}
```

Comment:
I think that this would be better as a more compact table. 

This isn't great but something like it that is smaller table might work better: 

```r
hash_table <-
  prep(ames_hash) %>% 
  bake(new_data = NULL, starts_with("Neighborhood_")) %>%
  bind_cols(ames_train %>% select(Neighborhood)) %>%
  distinct() %>%
  select(-Neighborhood) %>%
  map_dbl(sum) %>%
  enframe(value = "number of neighborhoods within a feature") %>% 
  count(`number of neighborhoods within a feature`, name = "number of occurrences")
```